src/platform/Linux/MdnsImpl.cpp

```diff
@@ -540,25 +541,40 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
     return browser == nullptr ? CHIP_ERROR_INTERNAL : CHIP_NO_ERROR;
 }
 
-MdnsServiceProtocol TruncateProtocolInType(char * type)
+MdnsServiceProtocol GetProtocolInType(const char * type)
 {
-    char * deliminator           = strrchr(type, '.');
-    MdnsServiceProtocol protocol = MdnsServiceProtocol::kMdnsProtocolUnknown;
+    const char * deliminator = strrchr(type, '.');
 
-    if (deliminator != NULL)
+    if (deliminator == NULL)
     {
-        if (strcmp("._tcp", deliminator) == 0)
-        {
-            protocol     = MdnsServiceProtocol::kMdnsProtocolTcp;
-            *deliminator = 0;
-        }
-        else if (strcmp("._udp", deliminator) == 0)
-        {
-            protocol     = MdnsServiceProtocol::kMdnsProtocolUdp;
-            *deliminator = 0;
-        }
+        ChipLogError(Discovery, "Failed to find protocol in type: %s", type);
+        return MdnsServiceProtocol::kMdnsProtocolUnknown;
+    }
+
+    if (strcmp("._tcp", deliminator) == 0)
+    {
+        return MdnsServiceProtocol::kMdnsProtocolTcp;
+    }
+    if (strcmp("._udp", deliminator) == 0)
+    {
+        return MdnsServiceProtocol::kMdnsProtocolUdp;
+    }
+
+    ChipLogError(Discovery, "Unknown protocol in type: %s", type);
+    return MdnsServiceProtocol::kMdnsProtocolUnknown;
+}
+
+template <size_t N>
+void CopyWithoutProtocol(char (&dest)[N], const char * typeAndProtocol)
+{
+    const char * dotPos          = strrchr(typeAndProtocol, '.');
+    size_t lengthWithoutProtocol = (dotPos != nullptr) ? static_cast<size_t>(dotPos - typeAndProtocol) : N;
+
+    Platform::CopyString(dest, typeAndProtocol);
+    if (lengthWithoutProtocol < N)
+    {
+        dest[lengthWithoutProtocol] = 0;
```

Comment:
Why must you still maintain the null-termination invariant yourself?

The CopyString will actually do the null-termination ehre even if lengthWithoutProtocol == N, due to the call `Platform::CopyString(dest, typeAndProtocol)` --> `Platform::CopyString(dest, typeAndProtocol,N)` internal expansion:

```
inline char * CopyString(char * dest, const char * source, size_t length)
{
    if (dest)
    {
        strncpy(dest, source, length);
        dest[length] = 0;
    }
    return dest;
}

/**
 * Convenience method for CopyString to auto-detect destination size.
 */
template <size_t N>
inline char * CopyString(char (&dest)[N], const char * source)
{
    return CopyString(dest, source, N);
}
```

Reply:
I need to truncate stuff after the .

Reply:
Updated with some renames and comments.

The CopyString method was buggy: it set a array[sizeof(array)] to 0 which is bad. Fixed that too - thank you for catching.

Reply:
Did a LOT of yak shaving here to make the CopyString seem a bit safer. I dislike the idea of 'pass in source length' and rather find 'pass in destination available size' to be a least unexpected behaviour.

Hoping code health is improved now. Also a bit frustrated that ChipMemStr has no unit tests.